RFC: Process Specific Working Directory (#144)
buildpacks/rfcs: text/0000-process-specific-working-directory.md

# Meta
[meta]: #meta
- Name: Process Specific Working Directory
- Start Date: 2021-03-09

> - Author(s): ForestEckhardt, rmoran

Reviewer: ```suggestion
- Author(s): ForestEckhardt, ryanmoran
```

Author: My bad

- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: N/A

# Summary
[summary]: #summary

For any given process, the buildpack should be able to specify a working
directory that is not `CNB_APP_DIR`.

# Motivation
[motivation]: #motivation

Allowing a buildpack to specify a custom working directory would allow
buildpacks to write start commands that need to be invoked from a specfic
directory without needing to invoke a shell.

# What it is
[what-it-is]: #what-it-is

It is a common pattern in buildpacks to execute the process from a directory
that is not `CNB_APP_DIR`. In these cases, buildpack authors are required to
write a process command that includes a change into the directory that the
process should run from. This often looks like `cd $CNB_APP_DIR/sub-directory &&
./execute-process`. Unfortunately, this means that the command can only be run
on a stack that includes a shell. It also means that including additional flags
or arguments to the command at runtime is cumbersome.

> As an alternative, the buildpack lifecycle could allow buildpacks to specify

Reviewer 2: Can you describe a bit more how this alternative could work?

the working directory of the process when it is invoked.

# How it Works
[how-it-works]: #how-it-works

A new field `directory` would be added the the `processes` in `launch.toml`.
```toml
[[processes]]
type = "<process type>"
command = "<command>"
args = ["<arguments>"]
direct = false

> directory = "<working directory>"

Reviewer 3: Is `path` more appropriate?

Author: To be honest with you, I wanted it to be `working-directory` but I went with `directory` for the sake of brevity. I don't know that `path` has the same connotation, `path` to me would be a field where I could put a separated list that would be searched for my command binary or what have you. Maybe that is just me though.

Reviewer 4: +1. `path` reads like `$PATH` to me. `working-directory` is clearest imo.

Reviewer 5: +1 to `working-dir` or `working-directory`

Author: I decided to opt for `working-directory`

```

> If `directory` is not set the launcher will be invoked in `CNB_APP_DIR` as

Reviewer 2: Summarizing from working group, might be good to clarify whether the change in working directory is intended to affect profile.d / exec.d for a given process.

Reviewer 2: Also it's worth noting that profile.d and exec.d can also exist __per process__ which might get a little confusing if the working directory for the process is not the same.

One idea that came up in implementation sub-team sync: only doing this for direct processes. (Since for non-direct processes, you could still `cd` into the directory at runtime).

Reviewer 4: @ForestEckhardt and I were chatting and we thought it'd be sort of confusing if the `working-directory` specified in the `launch.toml` impacted the working dir for profile.d and exec.d scripts. Reading the `launch.toml`, for me the most intuitive interpretation is that the `working-directory` applies to the `command` in the way that `args` also applies to the `command`. (Notably, `args` don't get passed to the profile.d or exec.d scripts.)

If a profile.d or exec.d scripts need to run in a certain working dir, they can `cd` in there at the start of their runs.

Reviewer 4: Though I haven't yet authored a buildpack that generates a profile.d or exec.d script. Would be interested to hear from buildpack authors about what makes sense/is convenient for them.

Reviewer: I'd agree with @fg-j's expectation for how `working-directory` would apply to a `profile.d` or `exec.d` script. In both cases they can be written to change their own working directory. We don't have the same kind of support for a process to change its working directory outside of invoking a shell command to do so.

normal. However, if given a relative path such as `src/app` the launcher should
be invoked from `$CNB_APP_DIR/src/app`. If given an absolute path such as
`/mounted/bin`, the launcher should invoke from `/mounted/bin`.

> 

Reviewer 5: What happens if the `directory` does not exist?

Author: I feel like you would get some `no such file or directory` error when trying to launch the image.

that is not `CNB_APP_DIR`. In these cases, buildpack authors are required to
write a process command that includes a change into the directory that the
process should run from. This often looks like `cd $CNB_APP_DIR/sub-directory &&
./execute-process`. Unfortunately, this means that the command can only be run
on a stack that includes a shell. It also means that including additional flags
or arguments to the command at runtime is cumbersome.

Here are some examples of where buildpacks have had to work around this issue:
* [Google Dotnet Publish](https://github.com/GoogleCloudPlatform/buildpacks/commit/a8d662f20cd3c304f0db4a82259400103e358429#diff-a56496a4700b75d6512c25eba9da9635ef937b1b38f50d6ce844275a77b09ac7R167)
* [Paketo NPM Start](https://github.com/paketo-buildpacks/npm-start/blob/d0e3f0a9375948c75dd663b56c95a0ee7b86556e/build.go#L41-L47)
* [Paketo Yarn Start](https://github.com/paketo-buildpacks/yarn-start/blob/d826b85e81df90f6d83e7bce1b581ba6bb79e6e7/build.go#L57)

As an alternative, the buildpack lifecycle could allow buildpacks to specify
the working directory of the process when it is invoked.

# How it Works
[how-it-works]: #how-it-works

> A new field `directory` would be added the the `processes` in `launch.toml`.

Reviewer 6: ```suggestion
A new field `working-directory` would be added the the `processes` in `launch.toml`.
```

```toml
[[processes]]
type = "<process type>"
command = "<command>"
args = ["<arguments>"]
direct = false
working-directory = "<working directory>"
```

> If `directory` is not set the launcher will be invoked in `CNB_APP_DIR` as

Reviewer 6: ```suggestion
If `working-directory` is not set the process will be invoked in `CNB_APP_DIR` as
```

> normal. However, if given a relative path such as `src/app` the launcher should

Reviewer 6: ```suggestion
normal. However, if given a relative path such as `src/app` the process should
```

be invoked from `$CNB_APP_DIR/src/app`. If given an absolute path such as

> `/mounted/bin`, the launcher should invoke from `/mounted/bin`.

Reviewer 6: ```suggestion
`/mounted/bin`, the process should invoke from `/mounted/bin`.
```

The `working-directory` specified for the process **should not** affect the working
directory for either `profile.d` or `exec.d` scripts.

The process specific `working-directory` **should not** affect the OCI image
metadata `WorkingDir` field.

# Prior Art
[prior-art]: #prior-art

- Go's [`exec.Cmd`](https://golang.org/pkg/os/exec/#Cmd) accepts a directory
  which that command will be executed in.

# Spec. Changes (OPTIONAL)
[spec-changes]: #spec-changes

> Add a new field `directory` field to `processes` in`launch.toml`

Reviewer 6: ```suggestion
Add a new field `working-directory` field to `processes` in`launch.toml`
```
